In [45]:
import pandas as pd
import numpy as np

df = pd.read_csv('datasets/1. projectfinal.csv')
d2 = pd.read_csv('datasets/1. projectfinal.csv')
print(d2)

       Unnamed: 0                    title     location  price(L)  \
0               0         Residential Plot  Maheshwaram     45.00   
1               1         Residential Plot     Kondakal     34.00   
2               2         Residential Plot    Bibinagar     14.19   
3               3         Residential Plot   Gachibowli    330.00   
4               4         Residential Plot    Shadnagar      6.72   
...           ...                      ...          ...       ...   
40255       40255          2 BHK Apartment  Chandanagar     47.00   
40256       40256  2 BHK Independent House  Kundanpally     25.74   
40257       40257          3 BHK Apartment  Peeramcheru     81.00   
40258       40258          2 BHK Apartment     Nizampet     34.50   
40259       40259          2 BHK Apartment  Chandanagar     40.00   

       rate_persqft  area_insqft     building_status  agent_rating  
0              1111         4050                 New           4.6  
1              2361         1440 

In [46]:
d3 = pd.read_csv('datasets/1. projectfinal.csv')

In [47]:
d2.shape

(40260, 8)

## Observation:

* The data we extracted contains 40260 rows and 8 columns same as our original data.


In [48]:
# Returns top 5 rows from data set(default value of head is 5).

d2.head()

Unnamed: 0             title     location  price(L)  rate_persqft  \
0           0  Residential Plot  Maheshwaram     45.00          1111   
1           1  Residential Plot     Kondakal     34.00          2361   
2           2  Residential Plot    Bibinagar     14.19           944   
3           3  Residential Plot   Gachibowli    330.00          3333   
4           4  Residential Plot    Shadnagar      6.72           466   

   area_insqft building_status  agent_rating  
0         4050             New           4.6  
1         1440             New           3.7  
2         1503             New           4.9  
3         9900             New           0.0  
4         1440             New           4.7

#### Observations

* Our features in the data set are title, location, price, rate_persqft, area_insqft, building_status, agent_rating

* **title**: Building Title ( 2BHK, 3BHK, Residential Plot etc)
* **location**: Location or Address where the building is located.
* **price**: Price of the building in Lakhs. Datatype Float.
* **rate_persqft**: Defines the price per square feet. of type int.
* **area_insqft**: Defines the size of the property in square feet which is of type int
* **building_status**the status of the building either new, old or construction in progress, of type Str
* **agent_rating**:  The rating given by the agent when he/she visited the property. of type float.


In [49]:
# Returns the aggregate functions for columns containing numerical data.

d2.describe()

Unnamed: 0      price(L)   rate_persqft    area_insqft  agent_rating
count  40260.000000  40260.000000   40260.000000   40260.000000  40260.000000
mean   20129.500000     65.331670    3297.212072    2360.068728      3.574724
std    11622.205255    126.247236    3850.078324    4190.331350      1.953369
min        0.000000      0.000000     100.000000     100.000000      0.000000
25%    10064.750000     20.000000    1018.000000    1290.000000      3.400000
50%    20129.500000     37.170000    2222.000000    1674.000000      4.600000
75%    30194.250000     78.630000    4800.000000    2250.000000      4.900000
max    40259.000000   7000.000000  172541.000000  215278.000000      5.000000

## Observations:

* Total Number of Rows: 40260
* Price:
    * Average: 65.33 Lakhs
    * std dev: 126.24
    * (min, max): (0, 7000)
    * Q1: 20
    * Q2: 37
    * Q3: 78


In [50]:
d2

Unnamed: 0                    title     location  price(L)  \
0               0         Residential Plot  Maheshwaram     45.00   
1               1         Residential Plot     Kondakal     34.00   
2               2         Residential Plot    Bibinagar     14.19   
3               3         Residential Plot   Gachibowli    330.00   
4               4         Residential Plot    Shadnagar      6.72   
...           ...                      ...          ...       ...   
40255       40255          2 BHK Apartment  Chandanagar     47.00   
40256       40256  2 BHK Independent House  Kundanpally     25.74   
40257       40257          3 BHK Apartment  Peeramcheru     81.00   
40258       40258          2 BHK Apartment     Nizampet     34.50   
40259       40259          2 BHK Apartment  Chandanagar     40.00   

       rate_persqft  area_insqft     building_status  agent_rating  
0              1111         4050                 New           4.6  
1              2361         1440                 New           3.7  
2               944         1503                 New           4.9  
3              3333         9900                 New           0.0  
4               466         1440                 New           4.7  
...             ...          ...                 ...           ...  
40255          4563         1030       Ready to move           0.0  
40256          3961          650       Ready to move           0.0  
40257          4500         1800       Ready to move           0.0  
40258          3833          900  Under Construction           0.0  
40259          4444          900  Under Construction           0.0  

[40260 rows x 8 columns]

In [51]:
d2['title'].value_counts()

Residential Plot            23726
3 BHK Apartment              7236
2 BHK Apartment              5001
2 BHK Independent House       958
3 BHK Villa                   648
4 BHK Villa                   491
4 BHK Apartment               490
1 BHK Apartment               396
3 BHK Independent House       278
1 BHK Independent House       219
4 BHK Independent House       209
2 BHK Villa                    88
2 BHK Independent Floor        80
5 BHK Apartment                71
5 BHK Independent House        56
5 BHK Villa                    55
6 BHK Independent House        50
3 BHK Independent Floor        45
7 BHK Independent House        26
1 BHK Independent Floor        22
4 BHK Independent Floor        19
10 BHK Independent House       17
1 RK Studio Apartment          11
6 BHK Apartment                10
9 BHK Independent House         9
8 BHK Independent House         7
1 BHK Villa                     6
6 BHK Independent Floor         5
6 BHK Villa                     4
10 BHK Apartme

In [52]:
d2.columns

Index(['Unnamed: 0', 'title', 'location', 'price(L)', 'rate_persqft',
       'area_insqft', 'building_status', 'agent_rating'],
      dtype='object')

In [53]:
d2['location'].value_counts()

Tellapur                  2896
Shadnagar                 2066
Gachibowli                1853
Kadthal                   1542
Maheshwaram               1521
                          ... 
Savarkar Nagar               1
Kalyanpuri Colony            1
Madhavi Nagar                1
Begumpet Main Road           1
Dwarkamai Nagar Colony       1
Name: location, Length: 935, dtype: int64

In [1]:
d2.duplicated().value_counts()

NameError: name 'd2' is not defined

In [55]:
d2.drop_duplicates()

Unnamed: 0                    title     location  price(L)  \
0               0         Residential Plot  Maheshwaram     45.00   
1               1         Residential Plot     Kondakal     34.00   
2               2         Residential Plot    Bibinagar     14.19   
3               3         Residential Plot   Gachibowli    330.00   
4               4         Residential Plot    Shadnagar      6.72   
...           ...                      ...          ...       ...   
40255       40255          2 BHK Apartment  Chandanagar     47.00   
40256       40256  2 BHK Independent House  Kundanpally     25.74   
40257       40257          3 BHK Apartment  Peeramcheru     81.00   
40258       40258          2 BHK Apartment     Nizampet     34.50   
40259       40259          2 BHK Apartment  Chandanagar     40.00   

       rate_persqft  area_insqft     building_status  agent_rating  
0              1111         4050                 New           4.6  
1              2361         1440                 New           3.7  
2               944         1503                 New           4.9  
3              3333         9900                 New           0.0  
4               466         1440                 New           4.7  
...             ...          ...                 ...           ...  
40255          4563         1030       Ready to move           0.0  
40256          3961          650       Ready to move           0.0  
40257          4500         1800       Ready to move           0.0  
40258          3833          900  Under Construction           0.0  
40259          4444          900  Under Construction           0.0  

[40260 rows x 8 columns]

In [56]:
d2.reset_index()

index  Unnamed: 0                    title     location  price(L)  \
0          0           0         Residential Plot  Maheshwaram     45.00   
1          1           1         Residential Plot     Kondakal     34.00   
2          2           2         Residential Plot    Bibinagar     14.19   
3          3           3         Residential Plot   Gachibowli    330.00   
4          4           4         Residential Plot    Shadnagar      6.72   
...      ...         ...                      ...          ...       ...   
40255  40255       40255          2 BHK Apartment  Chandanagar     47.00   
40256  40256       40256  2 BHK Independent House  Kundanpally     25.74   
40257  40257       40257          3 BHK Apartment  Peeramcheru     81.00   
40258  40258       40258          2 BHK Apartment     Nizampet     34.50   
40259  40259       40259          2 BHK Apartment  Chandanagar     40.00   

       rate_persqft  area_insqft     building_status  agent_rating  
0              1111         4050                 New           4.6  
1              2361         1440                 New           3.7  
2               944         1503                 New           4.9  
3              3333         9900                 New           0.0  
4               466         1440                 New           4.7  
...             ...          ...                 ...           ...  
40255          4563         1030       Ready to move           0.0  
40256          3961          650       Ready to move           0.0  
40257          4500         1800       Ready to move           0.0  
40258          3833          900  Under Construction           0.0  
40259          4444          900  Under Construction           0.0  

[40260 rows x 9 columns]

In [57]:
d2

Unnamed: 0                    title     location  price(L)  \
0               0         Residential Plot  Maheshwaram     45.00   
1               1         Residential Plot     Kondakal     34.00   
2               2         Residential Plot    Bibinagar     14.19   
3               3         Residential Plot   Gachibowli    330.00   
4               4         Residential Plot    Shadnagar      6.72   
...           ...                      ...          ...       ...   
40255       40255          2 BHK Apartment  Chandanagar     47.00   
40256       40256  2 BHK Independent House  Kundanpally     25.74   
40257       40257          3 BHK Apartment  Peeramcheru     81.00   
40258       40258          2 BHK Apartment     Nizampet     34.50   
40259       40259          2 BHK Apartment  Chandanagar     40.00   

       rate_persqft  area_insqft     building_status  agent_rating  
0              1111         4050                 New           4.6  
1              2361         1440                 New           3.7  
2               944         1503                 New           4.9  
3              3333         9900                 New           0.0  
4               466         1440                 New           4.7  
...             ...          ...                 ...           ...  
40255          4563         1030       Ready to move           0.0  
40256          3961          650       Ready to move           0.0  
40257          4500         1800       Ready to move           0.0  
40258          3833          900  Under Construction           0.0  
40259          4444          900  Under Construction           0.0  

[40260 rows x 8 columns]

In [58]:
d2.isnull().any()

Unnamed: 0         False
title              False
location           False
price(L)           False
rate_persqft       False
area_insqft        False
building_status    False
agent_rating       False
dtype: bool

In [59]:
d2.notnull().any()

Unnamed: 0         True
title              True
location           True
price(L)           True
rate_persqft       True
area_insqft        True
building_status    True
agent_rating       True
dtype: bool

In [60]:
d2.drop('Unnamed: 0', axis = 1, inplace = True)

In [61]:
d2.drop_duplicates(inplace=True)

In [62]:
d2.shape

(25715, 7)

In [63]:
d2[d2['price(L)'] == 0]

title                location  price(L)  rate_persqft  \
12768  Residential Plot         Vanasthalipuram       0.0           100   
15982  Residential Plot           Jubilee Hills       0.0           333   
22109  Residential Plot              Kukatpally       0.0           213   
22617  Residential Plot  Nagole Bandlaguda Road       0.0           106   

       area_insqft building_status  agent_rating  
12768          450             New           4.4  
15982          150             New           2.9  
22109          150             New           2.4  
22617          150             New           2.0

###  Observations

1. The price for few records is zero and in the business this can't be zero. So either we can drop them or treat them.
2. Treating the price: We need to collect the average price of a particular location(where the price is 0.0) and the square feet.

In [64]:
# Returns index positions of those properties whose price is 0.0L

n = d2[d2['price(L)'] == 0].index.tolist()
print(n)
for row in n:
    print("(", d2.loc[row, 'rate_persqft'], ", ", d2.loc[row, 'area_insqft'], " )" )
    

[12768, 15982, 22109, 22617]
( 100 ,  450  )
( 333 ,  150  )
( 213 ,  150  )
( 106 ,  150  )


In [65]:
# This code multiplies rate per sqft and area and returns the price for the properties
import numpy as np

rows = d2[d2['price(L)'] == 0].index.tolist()
prices = []

for row in rows:
    ratepersqft = d2.loc[row, 'rate_persqft'] # Return sthe value of ratepersqft.
    areainsqft = d2.loc[row, 'area_insqft']
    price = np.round(ratepersqft * areainsqft / 100000, 2)
    d2.loc[row, 'price(L)'] = price
    prices.append(price)
    

In [66]:
# Successfully Replaced
d2[d2['price(L)'] < 1]

title                location  price(L)  rate_persqft  \
12768  Residential Plot         Vanasthalipuram      0.45           100   
15982  Residential Plot           Jubilee Hills      0.50           333   
22109  Residential Plot              Kukatpally      0.32           213   
22617  Residential Plot  Nagole Bandlaguda Road      0.16           106   

       area_insqft building_status  agent_rating  
12768          450             New           4.4  
15982          150             New           2.9  
22109          150             New           2.4  
22617          150             New           2.0

In [67]:
prices

[0.45, 0.5, 0.32, 0.16]

In [68]:
d2[d2['price(L)'] < 1]

title                location  price(L)  rate_persqft  \
12768  Residential Plot         Vanasthalipuram      0.45           100   
15982  Residential Plot           Jubilee Hills      0.50           333   
22109  Residential Plot              Kukatpally      0.32           213   
22617  Residential Plot  Nagole Bandlaguda Road      0.16           106   

       area_insqft building_status  agent_rating  
12768          450             New           4.4  
15982          150             New           2.9  
22109          150             New           2.4  
22617          150             New           2.0

In [69]:
d2.loc[rows]

title                location  price(L)  rate_persqft  \
12768  Residential Plot         Vanasthalipuram      0.45           100   
15982  Residential Plot           Jubilee Hills      0.50           333   
22109  Residential Plot              Kukatpally      0.32           213   
22617  Residential Plot  Nagole Bandlaguda Road      0.16           106   

       area_insqft building_status  agent_rating  
12768          450             New           4.4  
15982          150             New           2.9  
22109          150             New           2.4  
22617          150             New           2.0

In [70]:
d2[d2['rate_persqft'] <= 0]

Empty DataFrame
Columns: [title, location, price(L), rate_persqft, area_insqft, building_status, agent_rating]
Index: []

In [71]:
d2[d2['area_insqft'] <= 0]

Empty DataFrame
Columns: [title, location, price(L), rate_persqft, area_insqft, building_status, agent_rating]
Index: []

In [72]:
d3.drop('Unnamed: 0', axis = 1, inplace = True)

#### Before Cleaning

In [73]:
d3.describe()

price(L)   rate_persqft    area_insqft  agent_rating
count  40260.000000   40260.000000   40260.000000  40260.000000
mean      65.331670    3297.212072    2360.068728      3.574724
std      126.247236    3850.078324    4190.331350      1.953369
min        0.000000     100.000000     100.000000      0.000000
25%       20.000000    1018.000000    1290.000000      3.400000
50%       37.170000    2222.000000    1674.000000      4.600000
75%       78.630000    4800.000000    2250.000000      4.900000
max     7000.000000  172541.000000  215278.000000      5.000000

### After Cleaning

In [74]:
d2.describe()

price(L)   rate_persqft    area_insqft  agent_rating
count  25715.000000   25715.000000   25715.000000  25715.000000
mean      68.963410    3491.056621    2328.315497      3.363387
std      140.980722    4218.740006    4836.964889      2.017042
min        0.160000     100.000000     100.000000      0.000000
25%       20.000000    1111.000000    1250.000000      0.000000
50%       38.980000    2556.000000    1645.000000      4.500000
75%       76.000000    4962.000000    2250.000000      4.800000
max     7000.000000  172541.000000  215278.000000      5.000000

In [75]:
d2

title     location  price(L)  rate_persqft  \
0             Residential Plot  Maheshwaram     45.00          1111   
1             Residential Plot     Kondakal     34.00          2361   
2             Residential Plot    Bibinagar     14.19           944   
3             Residential Plot   Gachibowli    330.00          3333   
4             Residential Plot    Shadnagar      6.72           466   
...                        ...          ...       ...           ...   
40255          2 BHK Apartment  Chandanagar     47.00          4563   
40256  2 BHK Independent House  Kundanpally     25.74          3961   
40257          3 BHK Apartment  Peeramcheru     81.00          4500   
40258          2 BHK Apartment     Nizampet     34.50          3833   
40259          2 BHK Apartment  Chandanagar     40.00          4444   

       area_insqft     building_status  agent_rating  
0             4050                 New           4.6  
1             1440                 New           3.7  
2             1503                 New           4.9  
3             9900                 New           0.0  
4             1440                 New           4.7  
...            ...                 ...           ...  
40255         1030       Ready to move           0.0  
40256          650       Ready to move           0.0  
40257         1800       Ready to move           0.0  
40258          900  Under Construction           0.0  
40259          900  Under Construction           0.0  

[25715 rows x 7 columns]

In [76]:
d2['location'].value_counts()

Shadnagar                         1448
Tellapur                          1252
Maheshwaram                       1025
Gachibowli                         891
Kadthal                            736
                                  ... 
Anjaneya Nagar                       1
Nagole Jaipur Colony Main Road       1
Journalist Colony                    1
Narayanguda                          1
Dwarkamai Nagar Colony               1
Name: location, Length: 935, dtype: int64

In [77]:
d2['location'].value_counts()

Shadnagar                         1448
Tellapur                          1252
Maheshwaram                       1025
Gachibowli                         891
Kadthal                            736
                                  ... 
Anjaneya Nagar                       1
Nagole Jaipur Colony Main Road       1
Journalist Colony                    1
Narayanguda                          1
Dwarkamai Nagar Colony               1
Name: location, Length: 935, dtype: int64

In [78]:
d2[d2['location'] == 'Shadnagar']['title'].value_counts()

Residential Plot           1435
4 BHK Villa                   5
5 BHK Villa                   2
2 BHK Independent House       2
1 BHK Independent House       1
3 BHK Independent Floor       1
3 BHK Villa                   1
2 BHK Villa                   1
Name: title, dtype: int64

In [101]:
sn_df = d2[d2['location'] == 'Shadnagar']
sn_df[sn_df['agent_rating'] == 5]

title   location  price(L)  rate_persqft  area_insqft  \
1596   Residential Plot  Shadnagar     11.70           866         1350   
1597   Residential Plot  Shadnagar     11.70          7800          150   
1598   Residential Plot  Shadnagar     11.25           833         1350   
1680   Residential Plot  Shadnagar     16.61          1255         1323   
1681   Residential Plot  Shadnagar     16.50          1247         1323   
...                 ...        ...       ...           ...          ...   
36242       4 BHK Villa  Shadnagar    390.00          6000         6500   
38226       4 BHK Villa  Shadnagar    420.00          6000         7000   
38227       4 BHK Villa  Shadnagar    360.00          6000         6000   
38228       5 BHK Villa  Shadnagar    480.00          6000         8000   
38229       5 BHK Villa  Shadnagar    450.00          6000         7500   

          building_status  agent_rating  
1596                  New           5.0  
1597                  New           5.0  
1598                  New           5.0  
1680                  New           5.0  
1681                  New           5.0  
...                   ...           ...  
36242  Under Construction           5.0  
38226  Under Construction           5.0  
38227  Under Construction           5.0  
38228  Under Construction           5.0  
38229  Under Construction           5.0  

[209 rows x 7 columns]

In [114]:
sn_df[(sn_df['agent_rating'] == 5) & (sn_df['price(L)'] < 45)].describe()

price(L)  rate_persqft  area_insqft  agent_rating
count  203.000000    203.000000   203.000000         203.0
mean    12.711527    835.561576  1643.586207           5.0
std      5.723487    579.981888   631.571995           0.0
min      1.290000    136.000000   150.000000           5.0
25%     10.320000    755.000000  1323.000000           5.0
50%     12.160000    856.000000  1386.000000           5.0
75%     15.770000   1010.500000  1800.000000           5.0
max     43.100000   7800.000000  5445.000000           5.0

In [111]:
d2

title     location  price(L)  rate_persqft  \
0             Residential Plot  Maheshwaram     45.00          1111   
1             Residential Plot     Kondakal     34.00          2361   
2             Residential Plot    Bibinagar     14.19           944   
3             Residential Plot   Gachibowli    330.00          3333   
4             Residential Plot    Shadnagar      6.72           466   
...                        ...          ...       ...           ...   
40255          2 BHK Apartment  Chandanagar     47.00          4563   
40256  2 BHK Independent House  Kundanpally     25.74          3961   
40257          3 BHK Apartment  Peeramcheru     81.00          4500   
40258          2 BHK Apartment     Nizampet     34.50          3833   
40259          2 BHK Apartment  Chandanagar     40.00          4444   

       area_insqft     building_status  agent_rating  
0             4050                 New           4.6  
1             1440                 New           3.7  
2             1503                 New           4.9  
3             9900                 New           0.0  
4             1440                 New           4.7  
...            ...                 ...           ...  
40255         1030       Ready to move           0.0  
40256          650       Ready to move           0.0  
40257         1800       Ready to move           0.0  
40258          900  Under Construction           0.0  
40259          900  Under Construction           0.0  

[25715 rows x 7 columns]

In [119]:
gach_df = d2[(d2['location'] == 'Gachibowli')]
gach_df

title    location  price(L)  rate_persqft  area_insqft  \
3      Residential Plot  Gachibowli    330.00          3333         9900   
46      3 BHK Apartment  Gachibowli    180.00          6141         2931   
60      3 BHK Apartment  Gachibowli    115.00          5400         2132   
61      3 BHK Apartment  Gachibowli    101.00          5400         1874   
62      3 BHK Apartment  Gachibowli     96.76          5399         1792   
...                 ...         ...       ...           ...          ...   
39906   3 BHK Apartment  Gachibowli     85.00          5862         1450   
39984   4 BHK Apartment  Gachibowli    170.00          7582         2242   
40018   3 BHK Apartment  Gachibowli     92.00          5569         1652   
40116   3 BHK Apartment  Gachibowli    110.00          5800         1900   
40246   3 BHK Apartment  Gachibowli    166.00          5995         2780   

          building_status  agent_rating  
3                     New           0.0  
46     Under Construction           0.0  
60     Under Construction           5.0  
61     Under Construction           5.0  
62     Under Construction           5.0  
...                   ...           ...  
39906       Ready to move           0.0  
39984       Ready to move           0.0  
40018       Ready to move           0.0  
40116  Under Construction           0.0  
40246  Under Construction           0.0  

[891 rows x 7 columns]

In [124]:
gach_df[gach_df['title'] == '4 BHK Apartment'].describe()

price(L)  rate_persqft  area_insqft  agent_rating
count    70.000000     70.000000    70.000000     70.000000
mean    267.485714   6924.814286  3816.557143      3.960000
std     131.976254   1940.386158   879.338546      1.746914
min     128.000000   3786.000000  2242.000000      0.000000
25%     198.000000   5641.000000  3380.000000      4.200000
50%     259.500000   6585.000000  3569.000000      4.800000
75%     307.250000   7885.500000  4120.750000      4.875000
max    1200.000000  18461.000000  6500.000000      5.000000

In [1]:
d2

NameError: name 'd2' is not defined